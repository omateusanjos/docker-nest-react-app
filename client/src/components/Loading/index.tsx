import React from "react";
import ReactLoading, { LoadingType } from "react-loading";

const Loading = ({
  type,
  color,
  ...rest
}: {
  type: LoadingType;
  color: string;
}) => (
  <ReactLoading
    type={type}
    color={color}
    height={"20%"}
    width={"20%"}
    delay={2}
    {...rest}
  />
);

export default Loading;
