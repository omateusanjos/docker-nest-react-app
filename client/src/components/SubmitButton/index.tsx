import React from "react";

interface SubmitButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  children: React.ReactNode;
}

const SubmitButton = ({ children, ...props }: SubmitButtonProps) => (
  <button
    className="text-logo-btn p-2 rounded-lg my-4 font-bold"
    type="submit"
    {...props}
  >
    {children}
  </button>
);

export { SubmitButton };
