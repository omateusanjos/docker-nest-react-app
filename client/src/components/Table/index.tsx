import React from "react";
import { useTable } from "react-table";
import { formatNumberToCurrency, Totalize } from "../../utils/table";

const Table = ({ columns, data }: { columns: any; data: any }) => {
  const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } =
    useTable({ columns, data });

  return (
    <>
      <table {...getTableProps()}>
        <thead>
          {headerGroups.map((headerGroup) => (
            <tr {...headerGroup.getHeaderGroupProps()}>
              {headerGroup.headers.map((column) => (
                <th {...column.getHeaderProps()}>{column.render("Header")}</th>
              ))}
            </tr>
          ))}
        </thead>

        <tbody {...getTableBodyProps()}>
          {rows.map((row) => {
            prepareRow(row);
            return (
              <tr {...row.getRowProps()}>
                {row.cells.map((cell) => {
                  const isValue = cell.column.Header?.toString() === "Value";

                  return (
                    <td {...cell.getCellProps()}>
                      {isValue
                        ? formatNumberToCurrency(cell.value)
                        : cell.render("Cell")}
                    </td>
                  );
                })}
              </tr>
            );
          })}
        </tbody>
      </table>
      <div>{formatNumberToCurrency(Totalize(data, "value"))}</div>
    </>
  );
};

export { Table };
