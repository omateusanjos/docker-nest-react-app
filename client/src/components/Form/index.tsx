import React, { useCallback, useContext } from "react";

import { useForm } from "react-hook-form";
import {
  checkFileAndReturnJSON,
  MockColumns,
  parserToTable,
} from "../../utils/form";

import { SubmitButton } from "../SubmitButton";

import { createTransactions } from "../../services/form";
import { FileContext } from "../../pages";
import useToast from "../../hooks/useToast";

const FileUploader = ({
  register,
  handleFileChange,
}: {
  register: ReturnType<typeof useForm>["register"];
  handleFileChange: (evt: React.ChangeEvent<HTMLInputElement>) => void;
}) => (
  <input
    type="file"
    {...register("arquivo")}
    onChange={handleFileChange}
    onChangeCapture={handleFileChange}
    onBlur={handleFileChange}
  />
);

export const Form = () => {
  const { ToastConfigs, ToastShow } = useToast();
  const { register, handleSubmit } = useForm();
  const { file, setContext } = useContext(FileContext);

  const handleFileChange = useCallback(
    (evt: React.ChangeEvent<HTMLInputElement>) => {
      if (!evt.currentTarget.files?.[0]) return false;
      checkFileAndReturnJSON(evt)?.then((data) => {
        const { columns, data: parsedData } = parserToTable(data);
        const isColumnsValid =
          JSON.stringify(columns) === JSON.stringify(MockColumns);
        if (!isColumnsValid) {
          console.log("Colunas inválidas");
          ToastShow({
            type: "error",
            message: "O arquivo enviado não é válido",
          });
          return;
        }

        setContext({
          columns,
          data: parsedData,
        });
      }) ?? setContext([]);
    },
    [setContext, ToastShow]
  );
  const onSucess = () => {
    ToastShow({ message: "Transações criadas com sucesso", type: "success" });
    setContext({
      columns: [],
      data: [],
      loading: false,
    });
  };

  const onError = () => {
    ToastShow({ message: "Erro ao registrar transações", type: "error" });
    setContext({ ...file, loading: false });
  };

  const submit = async () => {
    setContext({ ...file, loading: true });
    const response = await createTransactions(file.data);
    if (response.status === 201) return onSucess();
    onError();
  };
  return (
    <>
      <form onSubmit={handleSubmit(submit)}>
        <FileUploader register={register} handleFileChange={handleFileChange} />
        <SubmitButton hidden id="submit-button">
          Enviar
        </SubmitButton>

        <button
          id="cancel-button"
          type="button"
          onClick={() => setContext([])}
          hidden
        >
          Cancelar
        </button>
      </form>
      <ToastConfigs />
    </>
  );
};
