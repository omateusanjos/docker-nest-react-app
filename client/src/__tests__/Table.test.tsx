import React from "react";
import { render, RenderResult } from "@testing-library/react";
import { Table } from "../components/Table";

const mockData = [
  {
    name: "John Doe",
    email: "teste@gmail.com",
    phone: "123456789",
  },
];

const mockColumns = [
  {
    Header: "Name",
    accessor: "name",
  },
  {
    Header: "Email",
    accessor: "email",
  },

  {
    Header: "Phone",
    accessor: "phone",
  },
];

describe("Table component", () => {
  let wrapper: RenderResult | null;

  beforeEach(() => {
    jest.resetModules();
    wrapper = render(<Table data={mockData} columns={mockColumns} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
    wrapper = null;
  });

  it("should render a table", () => {
    expect(wrapper?.getByRole("table")).toBeInTheDocument();
  });

  it("should render a table with the correct number of rows", () => {
    expect(wrapper?.getAllByRole("row")).toHaveLength(2);
  });

  it("should render a table with the correct number of columns", () => {
    expect(wrapper?.getAllByRole("columnheader")).toHaveLength(3);
  });

  it("should render a table with the correct number of cells", () => {
    expect(wrapper?.getAllByRole("cell")).toHaveLength(3);
  });

  it("should render a table with the correct text", () => {
    expect(wrapper?.getByText("John Doe")).toBeInTheDocument();
    expect(wrapper?.getByText("teste@gmail.com")).toBeInTheDocument();
    expect(wrapper?.getByText("123456789")).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    expect(wrapper?.container).toMatchSnapshot();
  });
});
