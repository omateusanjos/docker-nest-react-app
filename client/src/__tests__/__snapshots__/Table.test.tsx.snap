// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component should match snapshot 1`] = `
<div>
  <table
    role="table"
  >
    <thead>
      <tr
        role="row"
      >
        <th
          colspan="1"
          role="columnheader"
        >
          Name
        </th>
        <th
          colspan="1"
          role="columnheader"
        >
          Email
        </th>
        <th
          colspan="1"
          role="columnheader"
        >
          Phone
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          John Doe
        </td>
        <td
          role="cell"
        >
          teste@gmail.com
        </td>
        <td
          role="cell"
        >
          123456789
        </td>
      </tr>
    </tbody>
  </table>
  <div>
    R$Â NaN
  </div>
</div>
`;
